### ISSUE TITLE
Suggestion: Use Make to build libraries/cores that include a Makefile

### ISSUE THREAD

**Issue by WestfW** [arduino/Arduino#2400]
As Arduino gets bigger and bigger, it starts to look like continuing to avoid the use of Make (or similar utility) might be a mistake.
As a transition and test mechanism, it might be relatively easy and painless to have the IDE scan source directories for a Makefile, and run Make if one is present (rather than iterating the build procedure for each known type of source file in the directory, as is done now.)

**Comment by lmihalkovic** [comment]
@mikaelpatel  thx for sharing your experience 

Reading the comments in this thread I (undoubtedly naively) don't see any technical issue standing in the way of an implementation (short of a lack of resources). Is anyone looking into it at the moment?!

**Comment by lmihalkovic** [comment]
I am looking toying with building support for arduino-makefile. At this point I isolated the builder behind a 
```SketchBuilder``

**Comment by lmihalkovic** [comment]
oops... lets try again... 

I am toying with building support for [arduino-makefile](https://github.com/sudar/Arduino-Makefile) inside the ide. At this point I isolated the builder behind a ```SketchBuilder``` interface for which I have 2 working implementations (see [4352](https://github.com/arduino/Arduino/issues/4352#issuecomment-171239954) ). 

As I don't want repeat the exercise for every kind of builder backend, I am isolating an API that should make it possible to have the builders maintained outside of the Arduino repo (basically anyone should be able to add support for theirs to the IDE). At the moment I am aiming for  each builder to define its own preferences page that will show up under /Compilation in the [preferences tree](https://github.com/arduino/Arduino/issues/4421). I will make it simple for each of these to display a table with some KEY/VALUE pairs that will then be exported into the proc env that will fork Make. I am also thinking about a config item to point to the location of the make executable.

Does anyone want to shoot bullets through this scheme?!

**Comment by lmihalkovic** [comment]
I got a few things done, but in the also I realized that the scheme is probably too restrictive as it was described. At this point I am going in the following direction:

each builder's preference page will have 
* a table for proc env NAME/VALUE pairs
* a table for cmd line -DNAME=VALUE pairs

If this is still too restrictive I am thinking about multiplexing them with a config-name table: for each name, the content of both tables can be used to drive make execution. In the end it means something like this in terms of variables

```
# conf1
CONF1.env.NAME1=VALUE1
CONF1.env.NAME2=VALUE2
CONF1.cmd.NAME1=VALUE1
CONF1.cmd.NAME2=VALUE2
# conf2
CONF2.env.NAME1=VALUE1
CONF2.env.NAME2=VALUE2
CONF2.cmd.NAME1=VALUE1
CONF2.cmd.NAME2=VALUE2
```

The core builder API is completely isolated such that it can be downloaded via Maven for anyone to write their own special builder backend.

Any though welcome.

**Comment by lmihalkovic** [comment]
easier to experiment with building Makefile based sketches first

<img width="400" alt="arduino-2016-01-20 at 12 33 30 pm" src="https://cloud.githubusercontent.com/assets/1318077/12467041/b6a30814-bfd8-11e5-9862-d01e262bb86b.png">

still have to understand more how [arduino-makefile](https://github.com/sudar/Arduino-Makefile) works.
 
@mikaelpatel  Cosa should be somewhat similar?!

**Comment by mikaelpatel** [comment]
@lmihalkovic 
I have added a layer of shell scripts and a default make rules on top of Arduino-Makefile. This removes the need to write a makefile. The boards is a parameter to the shell script. The default rules (Cosa.mk) set up necessary paths etc. 

There are three basic shell scripts in the Cosa build support. 1) build a sketch, 2) build core for all supported boards, 3) build all example sketches for given board. Shell script 2) and 3) are mainly for smoke-testing build targets. 

Bottom-line; it is possible to remove the need for makefiles but still have a make based build system.

**Comment by techpaul** [comment]
All this may be possible but I would prefer the final hex file (to upload) to be somewhere near the sketch so any time you need to programme another board or two it is not rebuild all because it was two days ago last built, but no code changes. Leaving even the final build file in temp areas is counter productive ion my view

**Comment by lmihalkovic** [comment]
@techpaul  agreed... I am seriously thinking about that one.

**Comment by techpaul** [comment]
@lmihalkovic having had a little thought I would suggest first time you build a core library for a board create a build directory (either near sketch or in program_data or var or wherever IDE/makefile area is as a core-<boad>.a Then rules of make or rebuild can be done as necessary. Not having to build for every board if you aint going to test compile or test on hardware for every type of board.

Possibly mimicing the user Arduino hierachy for sketches and contributed libraries examples, with subfolders for map and final build of each sketch/example.

Basically save any built libraries map and hex build files somewhere sane not temp.

**Comment by sandeepmistry** [comment]
I'm closing this as won't fix, we have arduino-cli now as an alternative.
